nmsg/kafkaio.c

```diff
@@ -587,31 +616,57 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
 	my_free(ctx);
 }
 
-static void
-_kafka_error_cb(rd_kafka_t *rk, int err, const char *reason, void *opaque)
-{
-	kafka_ctx_t ctx = (kafka_ctx_t) opaque;
-	rd_kafka_resp_err_t err_kafka = (rd_kafka_resp_err_t) err;
-	if (ctx == NULL) {
-		_nmsg_dprintf(2, "%s: unexpected Kafka opaque is NULL", __func__);
-		return;
-	}
-	switch (err_kafka) {
+static int
+_kafka_handle_error(rd_kafka_t *rk, kafka_ctx_t ctx, rd_kafka_resp_err_t err) {
+	int level = 2;
+
+	switch (err) {
 		/* Keep retrying on socket disconnect, brokers down and message timeout */
 		case RD_KAFKA_RESP_ERR__TRANSPORT:
 		case RD_KAFKA_RESP_ERR__ALL_BROKERS_DOWN:
+			if (ctx->connected)
+				_kafka_set_connected(ctx, __func__, false);
+			break;
 		case RD_KAFKA_RESP_ERR__MSG_TIMED_OUT:
-			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
+			ctx->timedout++;
+			ctx->timedout_total++;
+			if (ctx->timedout > 0 && (ctx->timedout % 100) == 0)
```

Comment:
First conditional is unnecessary; `ctx->timedout` will always be `> 0`

Reply:
Fixed